core/domain/html_validation_service.py

```diff
@@ -0,0 +1,767 @@
+
+    Returns:
+        bool. Boolean indicating whether a html string is valid for given RTE.
+    """
+    if rte_format == feconf.RTE_FORMAT_TEXTANGULAR:
+        RTE_TYPE = 'RTE_TYPE_TEXTANGULAR'
+    else:
+        RTE_TYPE = 'RTE_TYPE_CKEDITOR'
+    allowed_parent_list = feconf.RTE_CONTENT_SPEC[
+        RTE_TYPE]['ALLOWED_PARENT_LIST']
+    allowed_tag_list = feconf.RTE_CONTENT_SPEC[RTE_TYPE]['ALLOWED_TAG_LIST']
+
+    is_invalid = False
+
+    # Text with no parent tag is also invalid.
+    for content in soup.contents:
+        if not content.name:
+            is_invalid = True
+
+    for tag in soup.findAll():
+        # Checking for tags not allowed in RTE.
+        if tag.name not in allowed_tag_list:
+            if 'invalidTags' in err_dict:
+                err_dict['invalidTags'].append(tag.name)
+            else:
+                err_dict['invalidTags'] = [tag.name]
+            is_invalid = True
+        # Checking for parent-child relation that are not
+        # allowed in RTE.
+        parent = tag.parent.name
+        if (tag.name in allowed_tag_list) and (
+                parent not in allowed_parent_list[tag.name]):
+            if tag.name in err_dict:
+                err_dict[tag.name].append(parent)
+            else:
+                err_dict[tag.name] = [parent]
+            is_invalid = True
+
+    return is_invalid
+
+
+def add_caption_attr_to_image(html_string):
+    """Adds caption attribute to all oppia-noninteractive-image tags.
+
+    Args:
+        html_string. str: HTML string in which the caption attribute is to be
+            added.
+
+    Returns:
+        str. Updated HTML string with the caption attribute for all
+            oppia-noninteractive-image tags.
+    """
+    soup = bs4.BeautifulSoup(
+        html_string.encode(encoding='utf-8'), 'html.parser')
+
+    for image in soup.findAll(name='oppia-noninteractive-image'):
+        attrs = image.attrs
+        if 'caption-with-value' not in attrs:
+            image['caption-with-value'] = escape_html(json.dumps(''))
+
+    return unicode(soup)
+
+
+def validate_customization_args(html_list):
+    """Validates customization arguments of Rich Text Components in a list of
+    html string.
+
+    Args:
+        html_list: list(str). List of html strings to be validated.
+
+    Returns:
+        dict: Dictionary of all the invalid customisation args where
+            key is a Rich Text Component and value is the invalid html string.
+    """
+    # Dictionary to hold html strings in which customization arguments
+    # are invalid.
+    err_dict = {}
+    RICH_TEXT_COMPONENTS = OPPIA_INLINE_COMPONENTS + OPPIA_BLOCK_COMPONENTS
+
+    tag_dict = {}
+    for html_string in html_list:
+        soup = bs4.BeautifulSoup(
+            html_string.encode(encoding='utf-8'), 'html.parser')
+
+        for tag_name in RICH_TEXT_COMPONENTS:
+            for tag in soup.findAll(name=tag_name):
+                tag_dict[tag] = html_string
```

Comment:
I am storing the html strings since they are needed to be stored in err_dict with the error message for the tag. 

Also if two tags are completely same, it is fine to store either of the strings they occur in for testing.

Reply:
But what if you stored the tag string instead (in err_dict)?

Reply:
We can do that as well but if the tag is the part of larger string like:
```
<p>this is a link<oppia-noninteractive-link text-with-value="" url-with-value=""></oppia-noninteractive-link></p>
```
then I wanted to have the complete string during migration to avoid any mistakes. I don't have any specific case of mistake which can pop out, but since we encountered many weird cases earlier too, I wanted to make sure that we have the complete data.